### PR TITLE
[WIP] Work Stealing and Eager Worker Requesting Mode

### PR DESCRIPTION
**Pull request by goliaro**
## Why are these changes needed?

These changes enable work stealing among workers that are associated with the same owner. They also introduce a new 'eager' worker requesting mode, which allows us to request new workers even when the task queues at a given owner are already empty. The purpose of these new workers is to steal tasks from the other overloaded workers.

## Checks

- [x ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: src/ray/protobuf/core_worker.proto [hunk 1]
```diff
@@ -71,6 +71,15 @@ message ReturnObject {
   int64 size = 6;
 }
 
+message StealWorkRequest {
+  int64 max_tasks_to_steal = 1;
+}
+
+message StealWorkReply {
+  int64 number_of_tasks_stolen = 1;
```
Comment (stephanie-wang) [inline src/ray/protobuf/core_worker.proto:79]: I don't think we need this field since we can just check the size of `tasks_stolen`.

File: src/ray/core_worker/transport/direct_task_transport.cc [hunk 2]
```diff
@@ -93,13 +102,168 @@ Status CoreWorkerDirectTaskSubmitter::SubmitTask(TaskSpecification task_spec) {
...
+
+    ReturnWorker(thief_addr, was_error);
+    return;
+  }
+
+  auto request = std::unique_ptr<rpc::StealWorkRequest>(new rpc::StealWorkRequest);
+  RAY_LOG(DEBUG)
+      << "Calling Steal Work RPC! Maximum number of tasks requested (to steal) = "
+      << victim->second.tasks_in_flight_;
+  request->set_max_tasks_to_steal(victim->second.tasks_in_flight_);
```
Comment (stephanie-wang) [inline src/ray/core_worker/transport/direct_task_transport.cc:210]: This flag seems useful in theory since stealing can be expensive. However, I don't think it makes sense to set it to the total number of tasks in flight to the victim. The victim should already be basically limited to that number since its queue size should be <= the tasks in flight, right? (Assuming the owner doesn't send many more tasks to the victim in between this point and when the victim receives the steal request)

**Review by stephanie-wang — COMMENTED** [review]
I didn't read through `direct_task_transport.cc` code since I left several suggestions to reduce the scope of this PR (remove the eager worker requesting) and to refactor that will probably change the code significantly. We should try to keep the changes very simple and contained (right now there is a lot of logic that has to check whether work stealing is enabled and it's a little hard to follow).

**Comment by stephanie-wang** [comment]
Could you also add a high-level description of the algorithm in the PR text? It would be nice to see a comparison of the flow diagram between the previous code and the new code, e.g., when are new workers requested?

I also had a question about when workers are returned if there are no tasks left to steal. Maybe this is already covered, but I didn't read the code in direct_task_transport.cc very carefully :) For example, what happens in the following case?
1. Submit two tasks to worker A.
2. Request worker B since there is a task to steal.
3. Worker B is granted. Try to steal from worker A
4. Worker A finishes both tasks, responds 0 to the steal request.

When do we return worker B in this case?

File: src/ray/protobuf/core_worker.proto [hunk 3]
```diff
@@ -71,6 +71,15 @@ message ReturnObject {
...
+  repeated TaskSpec tasks_stolen = 2;
```
Comment (ericl) [inline src/ray/protobuf/core_worker.proto:80]: Doesn't the owner already know which specs are on the worker? No need to send the full spec message back, can just send their indices.

File: src/ray/protobuf/core_worker.proto [hunk 4]
```diff
@@ -71,6 +71,15 @@ message ReturnObject {
...
+}
+
 message PushTaskRequest {
```
Comment (ericl) [inline src/ray/protobuf/core_worker.proto:83]: To take full advantage of vectorization, don't we need to push a list of tasks to the worker instead of one per RPC?

File: src/ray/protobuf/core_worker.proto (hunk 3 above)
Comment (stephanie-wang) [reply src/ray/protobuf/core_worker.proto:80]: Hmm yeah, good point. It's probably best to send back the task ID instead of the indices since the indices could get updated between when the worker replies and when the owner receives the reply, though.

File: src/ray/protobuf/core_worker.proto (hunk 4 above)
Comment (stephanie-wang) [reply src/ray/protobuf/core_worker.proto:83]: We're still getting some improvement from grpc batching, but yeah, long-term we should consider sending multiple tasks to a worker in a single RPC. Right now it's a little tricky to do that without adding latency for other tasks because the return values of the task are also sent back in the reply.

Comment (ericl) [reply src/ray/protobuf/core_worker.proto:83]: Makes sense. You could do something like best effort vectorization of replies (send replies back along a separate reply path). This is probably something that would make sense once we have gRPC streaming support.

**Comment by stephanie-wang** [comment]
@gabrieleoliaro, let me know when this is ready for another look! :)

File: src/ray/common/ray_config_def.h [hunk 5]
```diff
@@ -328,6 +328,14 @@ RAY_CONFIG(bool, put_small_object_in_memory_store, false)
 /// pipelining task submission.
 RAY_CONFIG(uint32_t, max_tasks_in_flight_per_worker, 1)
 
+/// Whether work stealing among existing workers should be enabled.
+RAY_CONFIG(bool, work_stealing_enabled, false)
+
+/// Whether the eager worker requesting mode should be enabled, so that
+/// we can request new workers for the purpose of stealing work from overloaded
+/// existing workers.
+RAY_CONFIG(bool, work_stealing_and_eager_workers_requesting_enabled, false)
```
[the remark this answers is not in the document]
Comment (gabrieleoliaro) [reply src/ray/common/ray_config_def.h:337]: After discussing this offline, we are keeping work stealing and eager worker requesting together, but we are adding only one boolean variable (instead of two) to the config files

File: src/ray/core_worker/core_worker.cc [hunk 6]
```diff
@@ -1811,6 +1821,51 @@ void CoreWorker::HandlePushTask(const rpc::PushTaskRequest &request,
...
+                                 rpc::SendReplyCallback send_reply_callback) {
+  RAY_LOG(DEBUG) << "Entering HandleStealWork!";
+  // get maximum number of tasks to steal
+  int max_tasks_to_steal = request.max_tasks_to_steal();
+  RAY_CHECK(max_tasks_to_steal >= 0);
+
+  {
+    absl::MutexLock lock(&direct_task_receiver_->mu_);
+
+    int half = direct_task_receiver_->tasks_received_.size() / 2;
```
[the remark this answers is not in the document]
Comment (gabrieleoliaro) [reply src/ray/core_worker/core_worker.cc:1835]: Just made the change from `int` to `size_t`. When it comes to the choice of stealing _half_ of the task (instead of a larger or smaller fraction of the total), the idea is that when a _thief_ initiates stealing, it has no tasks in flight. So by stealing half of the tasks, we make sure that _thief_ and _victim_ end up with the same number of tasks in flight. Of course, if tasks are very short, when the _thief_ receives the stolen tasks, the _victim_ might have less than half of the original workload (because it has executed some in the meantime).

File: src/ray/core_worker/core_worker.cc [hunk 7]
```diff
@@ -1811,6 +1821,51 @@ void CoreWorker::HandlePushTask(const rpc::PushTaskRequest &request,
...
+
+    if (half == 0) {
+      RAY_LOG(DEBUG) << "We don't have enough tasks to steal, so we return early!";
+      reply->set_number_of_tasks_stolen(0);
+      send_reply_callback(Status::OK(), nullptr, nullptr);
+      return;
+    }
+
+    absl::flat_hash_map<TaskID, TaskSpecification>::iterator it =
+        direct_task_receiver_->tasks_received_.begin();
```
[the remark this answers is not in the document]
Comment (gabrieleoliaro) [reply src/ray/core_worker/core_worker.cc:1846]: That's a really good point. Let me change this now!

File: src/ray/protobuf/core_worker.proto (hunk 3 above)
Comment (gabrieleoliaro) [reply src/ray/protobuf/core_worker.proto:80]: Currently, I don't think the owner keeps a {`TaskId` -> `TaskSpecification`} mapping for the tasks that are in-flight to a worker. In fact, right after calling `PushNormalTask` to push a task to a worker, `OnWorkerIdle` pops the task's TaskSpec from the relevant owner's queue. So once we push a task, the owner will not see the task's TaskSpec until the worker responds to the owner and activates the `PushNormalTask` callback (which captures the TaskSpec).

For this reason, when we perform work stealing, the only way for the owner to access the the stolen tasks' TaskSpecs is for the victim to send them back along with the `StealWork` rpc reply. 

The worker will eventually also respond to each of the stolen tasks' original `PushNormalTask` RPCs, so an alternative approach would be to let those `PushNormalTask` callbacks take care of pushing the stolen tasks to the thief. However, this approach would be less efficient, because the replies would only start coming back after the worker has finished executing the non-stolen tasks that were pushed to the worker's queue before the stolen ones.

File: src/ray/core_worker/core_worker.cc [hunk 8]
```diff
@@ -1803,6 +1804,15 @@ void CoreWorker::HandlePushTask(const rpc::PushTaskRequest &request,
   }
 
   task_queue_length_ += 1;
+
+  const TaskSpecification task_spec(request.task_spec());
+  {
+    absl::MutexLock lock(&direct_task_receiver_->mu_);
+    RAY_LOG(DEBUG) << "Received task " << task_spec.TaskId()
+                   << ". Adding it to the tasks_received_ queue!";
+    direct_task_receiver_->tasks_received_.emplace(task_spec.TaskId(), task_spec);
```
[the remark this answers is not in the document]
Comment (gabrieleoliaro) [reply src/ray/core_worker/core_worker.cc:1813]: > Could you also add a high-level description of the algorithm in the PR text? It would be nice to see a comparison of the flow diagram between the previous code and the new code, e.g., when are new workers requested?
> 
> I also had a question about when workers are returned if there are no tasks left to steal. Maybe this is already covered, but I didn't read the code in direct_task_transport.cc very carefully :) For example, what happens in the following case?
> 
> 1. Submit two tasks to worker A.
> 2. Request worker B since there is a task to steal.
> 3. Worker B is granted. Try to steal from worker A
> 4. Worker A finishes both tasks, responds 0 to the steal request.
> 
> When do we return worker B in this case?

Sure!

In the situation you mention here, worker B's `StealWork` callback, upon receiving a negative response from worker A (the victim), would call the `ReturnWorker` function to return worker B to the Raylet.

**Comment by gabrieleoliaro** [comment]
> Could you also add a high-level description of the algorithm in the PR text? It would be nice to see a comparison of the flow diagram between the previous code and the new code, e.g., when are new workers requested?
> 
> I also had a question about when workers are returned if there are no tasks left to steal. Maybe this is already covered, but I didn't read the code in direct_task_transport.cc very carefully :) For example, what happens in the following case?
> 
> 1. Submit two tasks to worker A.
> 2. Request worker B since there is a task to steal.
> 3. Worker B is granted. Try to steal from worker A
> 4. Worker A finishes both tasks, responds 0 to the steal request.
> 
> When do we return worker B in this case?

Sure!

In the situation you mention here, worker B's StealWork callback, upon receiving a negative response from worker A (the victim), would call the ReturnWorker function to return worker B to the Raylet.

File: src/ray/core_worker/transport/direct_task_transport.h [hunk 9]
```diff
@@ -177,25 +202,32 @@ class CoreWorkerDirectTaskSubmitter {
...
   /// (2) The expiration time of a worker's lease.
   /// (3) The number of tasks that are currently in flight to the worker
+  /// (4) The current scheduling key of the tasks submitted to the worker
+  /// (5) The set of TaskIDs of the tasks that can be stolen by other workers
   struct LeaseEntry {
     std::shared_ptr<WorkerLeaseInterface> lease_client_;
     int64_t lease_expiration_time_;
     uint32_t tasks_in_flight_;
+    SchedulingKey current_scheduling_key_;
+    absl::flat_hash_set<TaskID> stealable_tasks_;
```
[the remark this answers is not in the document]
Comment (gabrieleoliaro) [reply src/ray/core_worker/transport/direct_task_transport.h:214]: The reason why we need to keep track of the number of stealable tasks as well as the number of tasks in flight (the former are a subset of the latter) is that the two quantities are often not equal, and using only one variable instead of two would cause some complications, inefficiencies, and perhaps errors.

When we steal a bunch of tasks from a worker, assuming that the reply to the thief's `StealWork` RPC arrives to the owner before the reply to the victim's `PushNormalTask` RPCs for the stolen tasks, we remove the stolen tasks from the `stealable_tasks_` hash_set, while the victim's `tasks_in_flight_` value initially remains unchanged. This reflects the fact that the stolen tasks are still technically in flight to the victim. Knowing how many tasks are in flight to a worker (even if they have been stolen) is important, for example, to know when to return the worker to the Raylet. So even if the stolen tasks are not going to be executed by the victim anymore, it is still useful for us to know that they are still "in flight", in the sense that their respective RPCs have not completed yet.

On the other hand, knowing the number of tasks in flight does not allow us, on its own, to perform work stealing in an efficient way. If we rely on it to choose which worker to steal from (picking the one with more tasks in flight), we will frequently make a suboptimal choice, as many times the number of tasks up for grab will be lower than the reported number of tasks in flight. 

Having decided to simultaneously keep track of the number of stealable tasks, as well as the number of tasks in flight to a worker, why use a `flat_hash_set` containing the TaskIDs of the `stealable_tasks_`, instead of just using an integer? The reason for this is to avoid a race condition between a thief's `StealWork` RPC callback and the victim's `PushNormalTask` RPC callbacks. When the thief steals some tasks from a victim, the reply to the `StealWork` RPC usually comes back to the owner earlier than the replies to the PushNormalTask RPCs. However, this might not always be the case. If the PushNormalTask callbacks are run first, they will be the ones responsible for decrementing the number of stealable tasks (to maintain the invariant that `stealable_tasks_.size() <= tasks_in_flight`). The challenge is that we must decrement the number of stealable tasks exactly once for each stolen task, and decrementing an integer is not an idempotent operation. On the other hand, deleting a `TaskID` from a set is idempotent, so we don't have to worry about finding a way to let the `PushNormalTask` and the `StealWork` callbacks coordinate to decrement the counter only once.

**Comment by gabrieleoliaro** [comment]
Closing this PR, superseded by PR [#10607](https://github.com/ray-project/ray/pull/10607)
